Add RealtimeConsumptionCatchupServiceCallback

npawar opened the pull request:
Creating a ServiceCallback within ServiceStatus, to check that realtime consuming segments have caught up. This will help with latency issues which can happen during server startup, because the server has to catchup on a lot of consumption.
For the initial version, this only adds a static wait time (configurable). More smarts can be added later (monitor stability or idleness of stream consumption rate)

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java:
```diff
@@ -129,4 +129,11 @@ public static boolean isTableResource(@Nonnull String resourceName) {
   }
+
+  /**
+   * Return whether the given resource name represents an realtime table resource.
```
nit: **a** realtime table resource

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
     }
   }
+  
+  public static abstract class BaseServiceStatusCallback implements ServiceStatusCallback {
```
I don't think we need this base class. ServiceStatusCallback does not always associate with Helix cluster info (e.g. MultipleCallbackServiceStatusCallback), also the methods are not very commonly used. I don't like the way of returning statusDescription by setting a member variable from the parent class, it's not very readable. The base class IdealStateMatchServiceStatusCallback is already good enough IMO.

npawar replied:
RealtimeConsumptionCatchupServiceCallback doesn't really belong under IdealStateMatchServiceStatusCallback. What is the suggestion? Keep RealtimeConsumptionCatchupServiceCallback independent and directly implement ServiceStatusCallback interface?

Jackie-Jiang replied:
Yes, directly implement ServiceStatusCallback

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+  public static class RealtimeConsumptionCatchupServiceStatusCallback extends BaseServiceStatusCallback {
+
+    private final Map<String, List<String>> _tableToConsumingSegmentsMap;
+    private long _endWaitTime = 0;
```
final?

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+              Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentName);
+              if (CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel.CONSUMING.equals(
+                  instanceStateMap.get(instanceName))) {
+                consumingSegments.add(segmentName);
```
I suggest to leave it as a boolean and break if we have even one consuming segment we own.

We can add new logic when we decide to examine each segment as we need to.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+
+      if (_tableToConsumingSegmentsMap.isEmpty()) {
+        LOGGER.info(
+            "No consuming segments to monitor on instance. Setting realtime consumption catchup wait time to 0ms");
```
Directly set the service status to GOOD instead of relying on the _endWaitTime

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+        LOGGER.info(
+            "No consuming segments to monitor on instance. Setting realtime consumption catchup wait time to 0ms");
+      } else {
+        long startTime = System.currentTimeMillis();
```
Should you take the time when this method is entered?

It may also be worth commenting that even in the not-so-fancy determination of  the time a consuming segment is ready to serve queries, the computation should be:
 end-time = (time of creation of partition consumer) + (confugured max time to catch up).

We are sort of approximating it to :+1: 
end-time = (time of server startup) + (configured max timeto catch up)

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+
+    @Override
+    public Status getServiceStatus() {
+      long now = System.currentTimeMillis();
```
Check for previous service status first, directly return GOOD if it is already GOOD. (Do not keep updating the statusDescription)

npawar replied:
Worked on all suggestions

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+      } else {
+        long startTime = System.currentTimeMillis();
+        _endWaitTime = startTime + TimeUnit.SECONDS.toMillis(realtimeConsumptionCatchupWaitTimeSeconds);
+        LOGGER.info("Monitoring realtime consumption catchup for tables:{}. Will wait for time:{} ms",
```
We are not really waiting here, so the log could be misleading. May be say "will allow until time {} before marking status as GOOD"?

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+    public Status getServiceStatus() {
+      long now = System.currentTimeMillis();
+      if (now < _endWaitTime) {
+        _statusDescription = String.format("Waiting for CONSUMING segments to catchup, timeRemaining=%dms", _endWaitTime - System.currentTimeMillis());
```
Use 'now' instead of calling System.currentMillis again

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -127,34 +129,129 @@ public String getStatusDescription() {
+        _statusDescription = String.format("Waiting for CONSUMING segments to catchup, timeRemaining=%dms", _endWaitTime - System.currentTimeMillis());
+        return Status.STARTING;
+      }
+      _statusDescription = "Wait time for CONSUMING segments catchup expired, now:" + now + ", endWaitTime:" + _endWaitTime;
```
I would omit the word 'wait' here (so we can grep for wait and see where waits are happening). How about: "Consuming segments caught up at " + _endWaitTime. 

Or, "Consuming segment status GOOD since " + _endTime.

npawar replied:
Fixed all suggestions. Please take a look @mcvsubbu

mcvsubbu commented:
lgtm, @Jackie-Jiang had some comments on class hierarchy.  I am neutral on that.

kishoreg commented:
@npawar  I have a few questions. 

What is the impact on the latency and when does it kick in and for what kind of use cases? 
How will the user know whats the right wait-time? We might still see the same issue if the server has not caught up within the wait time right?

One of the concerns here is that we stop the entire server from serving any other segment.

Looks like we need two things
- limit the number of threads for consumption so that it does not hog all the cpu during bootstrap/server startup
-  Have the logic in the state transition for OFFLINE -> CONSUMING to not ack Helix until it's caught up. This has the added benefit that query results are up to date across all replicas.

I am ok with the temporary fix but was wondering what would be the right fix that requires minimal configuration from the user.

npawar commented:
> @npawar I have a few questions.
> 
> What is the impact on the latency and when does it kick in and for what kind of use cases?
> How will the user know whats the right wait-time? We might still see the same issue if the server has not caught up within the wait time right?
> 
> One of the concerns here is that we stop the entire server from serving any other segment.
> 
> Looks like we need two things
> 
> * limit the number of threads for consumption so that it does not hog all the cpu during bootstrap/server startup
> * Have the logic in the state transition for OFFLINE -> CONSUMING to not ack Helix until it's caught up. This has the added benefit that query results are up to date across all replicas.
> 
> I am ok with the temporary fix but was wondering what would be the right fix that requires minimal configuration from the user.

The impact we've observed is usually for high ingestion rate realtime tables. The tables can usually take 2-3 minutes to completely catchup after server restart. During this time, we not only see high latencies/slow queries/timeouts because of cpu bottleneck, but also the results returned by that server will not be in sync with what other replicas return. 

We decided to have a simple approach of static wait time to begin with, until we can get more information about the consumption rate. Yes we will still see the issue if the server has not caught up in the configured time. 
We plan to include consumption rate related stats in server stats. Once we have those, we will remove the static wait. We will then wait until consumption rate stabilizes to the average rate or until idle.

I can make the static wait to default 0 if having the default 5 is a problem?

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -174,6 +174,8 @@ public ServerType getServerType() {
     public static final String CONFIG_OF_STARTER_ENABLE_SEGMENTS_LOADING_CHECK =
         "pinot.server.starter.enableSegmentsLoadingCheck";
     public static final String CONFIG_OF_STARTER_TIMEOUT_IN_SECONDS = "pinot.server.starter.timeoutInSeconds";
+    public static final String CONFIG_OF_STARTER_REALTIME_CONSUMPTION_CATCHUP_WAIT_TIME_SECONDS =
```
I would put this along with the **CONFIG_OF_SERVER_MIN_RESOURCE_PERCENT_FOR_START** since both of them are for the status check instead of the starter. Maybe rename it to "pinot.server.startup.realtimeConsumptionCatchupTimeMs"?

kishoreg commented:
Agree with providing a simple wait for now but let's keep the default to 0 and allow users to override if needed.

sunithabeeram commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -128,11 +129,91 @@ public String getStatusDescription() {
+    private final long _endWaitTime;
+
+    /**
+     * Realtime consumption catchup service which adds a static wait time for
```
Comment appears to be incomplete

sunithabeeram commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -128,11 +129,91 @@ public String getStatusDescription() {
+    }
+
+    @Override
+    public Status getServiceStatus() {
```
Shouldn't this be synchronized?

npawar commented:
Fixed suggestions @sunithabeeram 

@mcvsubbu I have refactored it such that we are not making the call to getResourcesInCluster and getResourceIdealState in every constructor. It is now happening only in 1 constructor.

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java:
```diff
@@ -129,56 +132,157 @@ public String getStatusDescription() {
-    private final int _minResourcesStartCount;
-
-    private String _statusDescription = STATUS_DESCRIPTION_INIT;
+    public IdealStateServiceStatusCallback(HelixManager helixManager, String clusterName, String instanceName,
```
So, this is a multi-service callback at the same time it is not? I am getting confused deciding one way or another.

Another approach could be that we provide a method in serviceStatus that gets the resources to be monitored, given the parameters. We have another constructor for the individual servicesStatus objects that also takes int he resources to be monitored, and HelixBrokerStarter (or server starter) invokes that constructor. I think that has two advantages:
1. It is better readable (or, at least I think so :)
2. We don't get all idealstates in the broker case. We only get the broker ideal state resource!

npawar replied:
i changed it, to make the HelixServerStarter and HelixBrokerStarter decide the resources to monitor. I don't like the idea of adding a method in ServiceStatus to fetch resources to monitor. Firstly, there's no neat way to distinguish between whether the request is being made by server/broker/test (other than check prefix of instance), and we'll end up with a lot of if else. Secondly, we need to return resourcesToMonitor, as well as boolean indicating need for Realtiem handler (in future, another list to hold consuming segments). 
lmk if current iteration works. @mcvsubbu

mcvsubbu commented on pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/HelixBrokerStarter.java:
```diff
@@ -237,16 +238,30 @@ public void start()
+  private void registerServiceStatusHandler() {
+    List<String> resourcesToMonitor = new ArrayList<>(1);
+    IdealState resourceIdealState = _helixAdmin.getResourceIdealState(_clusterName, Helix.BROKER_RESOURCE_INSTANCE);
+    if (resourceIdealState != null && resourceIdealState.isEnabled()) {
```
dont we also need to check if this broker instance is on for any of the partitions?

npawar replied:
Hmm yes, fixed. 
Follow up question: We still want to set the ServiceStatusHandlers, even if resourcesToMonitor is empty right? That way the health checks will get response GOOD instead of having no service handlers?

mcvsubbu replied:
If each service handler we want to have end up being empty, then yes. we do need to set one that always returns GOOD. Perhaps that can be the default setting for ServiceStatus.getServiceStatus ?
